RFC21: add new PRIORITY state

garlick opened the pull request:
As discussed in flux-framework/flux-core#3311, this PR adds a new PRIORITY job state that a job passes through when it is assigned its initial priority by a job manager priority plugin.

This should make integration of a priority plugin a bit easier, and also provide some user visibility in case the plugin takes a long time to assign the initial priority (for example if it has to fetch data from flux-accounting before it can answer).

Events normally drive job state transitions.  I added a `prioritize` event for this purpose, but footnoted it with the caveat that, unlike the other primary events, it will not be posted to the job eventlog.   This is because we surely do not want to log the repeated priority adjustments of a fair share priority plugin for posterity, and a fresh priority can easily be recalculated if the instance restarts with jobs in SCHED state.

That does mean that a job can regress from SCHED to PRIORITY state on an instance restart.   It does not mean that the job repeatedly does that as the priority is updated, since the job may have an outstanding alloc request to the scheduler that we do not want to cancel every time the priority is updated.  (Updating the scheduler's priority info will be the subject of another PR to RFC 27)

garlick commented:
Oh man, no idea what this means.  Readthedocs is working fine on my laptop (`make html`).  I can't tell if this even involves the rfc I changed.  Not?
```
Successfully built pyyaml pyrsistent
ERROR: Exception:
Traceback (most recent call last):
  File "/home/docs/checkouts/readthedocs.org/user_builds/flux-rfc/envs/271/lib/python3.7/site-packages/pip/_internal/cli/base_command.py", line 228, in _main
    status = self.run(options, args)
  File "/home/docs/checkouts/readthedocs.org/user_builds/flux-rfc/envs/271/lib/python3.7/site-packages/pip/_internal/cli/req_command.py", line 182, in wrapper
    return func(self, options, args)
  File "/home/docs/checkouts/readthedocs.org/user_builds/flux-rfc/envs/271/lib/python3.7/site-packages/pip/_internal/commands/install.py", line 379, in run
    requirement_set
  File "/home/docs/checkouts/readthedocs.org/user_builds/flux-rfc/envs/271/lib/python3.7/site-packages/pip/_internal/resolution/resolvelib/resolver.py", line 191, in get_installation_order
    weights = get_topological_weights(graph)
  File "/home/docs/checkouts/readthedocs.org/user_builds/flux-rfc/envs/271/lib/python3.7/site-packages/pip/_internal/resolution/resolvelib/resolver.py", line 242, in get_topological_weights
    assert len(weights) == len(graph)
AssertionErrorr
```

SteVwonder commented:
Looks like we smacked into a pip dependency resolution bug: https://github.com/pypa/pip/issues/9031

I don't quite grok their proposed solution of removing Sphinx as a requirement.  I'm guess they mean Sphinx is already included as an intermediate/implicit dependency thanks to sphinx-rtd-theme and/or sphinxcontrib-spelling.  So maybe dropping [sphinx from requirements.txt](https://github.com/flux-framework/rfc/blob/master/requirements.txt#L1) will work for us 🤷 .

Unfortunately, I don't see a way to have RTD not use the 2020 resolver 🤦

SteVwonder commented:
@garlick : if you run `pip install --upgrade pip` and then `pip install --use-feature 2020-resolver --exists-action=w --no-cache-dir -r requirements.txt`, it might reproduce locally

garlick commented:
Thanks for that hint.  It didn't reproduce but for some reason some extraneous warnings went away and I now noticed that one of the JSON examples is called out as invalid JSON and could not be syntax highlighted.  I tacked on a commit to fix that.

SteVwonder commented:
> t didn't reproduce 

Darn. The only thing that I can think to try currently is to drop the explicit `sphinx` from the requirements.txt and let it just be implicitly picked up as a dependency from the other packages.  That seemed to work for the other people in thread.

garlick commented:
Ah thanks.  Well that seems to have worked.  Are we OK with that change getting merged or does this need more thought?

On re-reviewing this proposed change, one other thing I should have pointed out for discussion is I defined the priority as a `double` with a range of 0-1.0.  My thought was that this value will become the _only_ value used to order the queue, with t_submit and administrative priority becoming factors that are input to a multi-factor priority plugin, probably with a default "builtin plugin" that _only_ considers those factors.

Would it be better to go with an integer value here?  how about the range?

grondo reviewed (commented):
Just caught one typo on the first pass through.

garlick commented on spec_21.rst:
```diff
@@ -190,15 +195,32 @@ Example:
+Prioritize Event
+^^^^^^^^^^^^^^^^
+
+Job has been (re-)assigned a priority by the job shell priority plugin:
```
I'll fix that.

Also, based on our discussions offline about the confusion of the PRIORITY state vs `priority` event that would affect it indirectly, not directly, I think I'll propose changing the `priority` event to `admin-priority` and `prioritize` to `priority`.

cmoussa1 commented:
> On re-reviewing this proposed change, one other thing I should have pointed out for discussion is I defined the priority as a `double` with a range of 0-1.0. My thought was that this value will become the only value used to order the queue, with t_submit and administrative priority becoming factors that are input to a multi-factor priority plugin, probably with a default "builtin plugin" that only considers those factors. Would it be better to go with an integer value here? how about the range?

If we were to model our multi-factor priority calculation in a similar manner to Slurm, I believe they define their job priority value as an `int` that ranges between 0 and 4294967295. [Here](https://slurm.schedmd.com/priority_multifactor.html) is where they define their formula. All of the _factors_ used in this calculation are floating point numbers that range from 0.0 to 1.0. I hope this helps!

chu11 commented on spec_21.rst:
```diff
@@ -195,6 +195,20 @@ Example:
    {"timestamp":1552593348.073045,"name":"submit","context":{"priority":16,"userid":5588,"flags":0}}
 
 
+Depend Event
```
commit message typo "mising" -> "missing"

chu11 commented:
> On re-reviewing this proposed change, one other thing I should have pointed out for discussion is I defined the priority as a double with a range of 0-1.0. My thought was that this value will become the only value used to order the queue, with t_submit and administrative priority becoming factors that are input to a multi-factor priority plugin, probably with a default "builtin plugin" that only considers those factors.

my first thought is that when priority plugin developers are debugging, it may be visually easier to see integers vs. floats.  Like I may not want to see "0.83738937" in logs all over the place.

garlick commented:
Thanks @cmoussa1!  An integer with range of 0 to 4294967295 sounds completely reasonable to me, and presumably it will make comparing slurm and flux priority algorithms slightly easier if we don't introduce arbitrary changes here.

garlick commented:
Forced a push with the priority redefined as an integer from 0 - 2^32-1, and fixed the spelling error @chu11 pointed out in my commit message. I also reworded that commit message, and added one to the main commit that adds the new state.

chu11 commented:
an aside, should we update all language in flux-core about priority to "administrative priority"?  e.g. `flux job priority`, output from `flux-jobs`, etc. 

On the one hand, consistency.  On the other hand, presumably at the user/admin level, "administrative priority" is the only priority they know, so it shouldn't matter.

garlick commented:
> On the other hand, presumably at the user/admin level, "administrative priority" is the only priority they know, so it shouldn't matter.

This is the way I was leaning.  Where it might get confusing is if we want to let users see what their current (caculated) priority value is.  What do we call that?

chu11 commented:
> Where it might get confusing is if we want to let users see what their current (caculated) priority value is. What do we call that?

queue priority?

garlick commented:
Ah sure, that makes sense.  So for example if we want flux job info to be able to list it in a custom formatted output, `{queue-priority}`.  I like it.

grondo commented:
The PRIORITY state should also be added to the list for the ACTIVE virtual state

grondo reviewed (approved):
LGTM! Thanks!

garlick commented:
Thanks, I'll set MWP
